// uu_appstatisticsg01/src/api/controllers/custom-stats-controller.js
"use strict";
//@@viewOn:revision
// coded: Eduard Batiushchenko (9108-9377-1)
// reviewed:
//@@viewOff:revision

const GetGeneralFileAttributesAbl = require("../../abl/custom-stats/get-general-file-attributes-abl");
const GetManagementFileAttributesAbl = require("../../abl/custom-stats/get-management-file-attributes-abl");
const GetExecutivesFileAttributesAbl = require("../../abl/custom-stats/get-executives-file-attributes-abl");
const PrivilegedGetExecutivesFileAttributesAbl = require("../../abl/custom-stats/privileged-get-executives-file-attributes-abl");
const PrivilegedGetManagementFileAttributesAbl = require("../../abl/custom-stats/privileged-get-management-file-attributes-abl");
const RecalculateCustomFileAbl = require("../../abl/custom-stats/recalculate-custom-file-abl");
const PrivilegedRecalculateCustomFileAbl = require("../../abl/custom-stats/privileged-recalculate-custom-file-abl");
const UpdateCustomFileAbl = require("../../abl/custom-stats/update-custom-file-abl");
const PrivilegedUpdateCustomFileAbl = require("../../abl/custom-stats/privileged-update-custom-file-abl");

class CustomStatsController {
  getGeneralFileAttributes(ucEnv) {
    return GetGeneralFileAttributesAbl.getGeneralFileAttributes(ucEnv.getUri().getAwid(), ucEnv.getDtoIn());
  }
  getManagementFileAttributes(ucEnv) {
    return GetManagementFileAttributesAbl.getManagementFileAttributes(ucEnv.getUri().getAwid(), ucEnv.getDtoIn());
  }
  getExecutivesFileAttributes(ucEnv) {
    return GetExecutivesFileAttributesAbl.getExecutivesFileAttributes(ucEnv.getUri().getAwid(), ucEnv.getDtoIn());
  }
  privilegedGetExecutivesFileAttributes(ucEnv) {
    return PrivilegedGetExecutivesFileAttributesAbl.privilegedGetExecutivesFileAttributes(
      ucEnv.getUri().getAwid(),
      ucEnv.getDtoIn()
    );
  }
  privilegedGetManagementFileAttributes(ucEnv) {
    return PrivilegedGetManagementFileAttributesAbl.privilegedGetManagementFileAttributes(
      ucEnv.getUri().getAwid(),
      ucEnv.getDtoIn()
    );
  }
  recalculateCustomFile(ucEnv) {
    return RecalculateCustomFileAbl.recalculateCustomFile(ucEnv.getUri().getAwid(), ucEnv.getDtoIn());
  }
  privilegedRecalculateCustomFile(ucEnv) {
    return PrivilegedRecalculateCustomFileAbl.privilegedRecalculateCustomFile(
      ucEnv.getUri().getAwid(),
      ucEnv.getDtoIn()
    );
  }
  updateCustomFile(ucEnv) {
    return UpdateCustomFileAbl.updateCustomFile(ucEnv.getUri().getAwid(), ucEnv.getDtoIn());
  }
  privilegedUpdateCustomFile(ucEnv) {
    return PrivilegedUpdateCustomFileAbl.privilegedUpdateCustomFile(ucEnv.getUri().getAwid(), ucEnv.getDtoIn());
  }
}

module.exports = new CustomStatsController();
