// uu_appstatisticsg01/src/api/warnings/custom-stats-warnings.js
"use strict";
//@@viewOn:revision
// coded: Eduard Batiushchenko (9108-9377-1)
// reviewed:
//@@viewOff:revision

const Errors = require("../errors/custom-stats-errors");

const WARNINGS = {
  getGeneralUnsupportedKeys: {
    code: `${Errors.GetGeneralFileAttributes.UC_CODE}unsupportedKeys`,
  },
  getManagementFileAttributes: {
    code: `${Errors.GetManagementFileAttributes.UC_CODE}unsupportedKeys`,
  },
  getExecutivesFileAttributes: {
    code: `${Errors.GetExecutivesFileAttributes.UC_CODE}unsupportedKeys`,
  },
  privilegedGetExecutivesFileAttributes: {
    code: `${Errors.PrivilegedGetExecutivesFileAttributes.UC_CODE}unsupportedKeys`,
  },
  privilegedGetManagementFileAttributes: {
    code: `${Errors.PrivilegedGetManagementFileAttributes.UC_CODE}unsupportedKeys`,
  },
  recalculateCustomFile: {
    code: `${Errors.RecalculateCustomFile.UC_CODE}unsupportedKeys`,
  },
  privilegedRecalculateCustomFile: {
    code: `${Errors.PrivilegedRecalculateCustomFile.UC_CODE}unsupportedKeys`,
  },
  updateCustomFile: {
    code: `${Errors.UpdateCustomFile.UC_CODE}unsupportedKeys`,
  },
  privilegedUpdateCustomFile: {
    code: `${Errors.PrivilegedUpdateCustomFile.UC_CODE}unsupportedKeys`,
  },
};

module.exports = WARNINGS;
