// uu_appstatisticsg01/src/abl/t-file/get-t-file-attributes-abl.js
// template: privileged-get-t-file-attributes-abl
// templateVersion: 0.1.1
//@@viewOn:revision
// coded: Hrynkiv Andriy (5890-4974-1)
// reviewed: Lukas Sykora (12-7774-1) - 30.03.2022
//@@viewOff:revision
//@@viewOn:constants

"use strict";
const { Validator } = require("uu_appg01_server").Validation;
const { ValidationHelper } = require("uu_appg01_server").AppServer;
const Errors = require("../../api/errors/t-files-error");
const WARNINGS = require("../../api/warnings/t-files-warnings");
const Constants = require("../../components/constants/t-file-constans");
const AppClient = require("uu_appg01_server").AppClient;
const { Config } = require("uu_appg01_server").Utils;
// eslint-disable-next-line no-unused-vars
const { getSystemIdentity, getAuthorizedProfiles } = require("../../helpers/app-stats-helper");

class GetTFileAttributesAbl {
  constructor() {
    this.validator = Validator.load();
  }

  async getTFileAttributes(dtoIn, authzContext) {
    //HDS 1
    let validationResult = this.validator.validate("uuAppstatisticsGetTFileAttributesDtoInTypes", dtoIn);
    let uuAppErrorMap = ValidationHelper.processValidationResult(
      dtoIn,
      validationResult,
      WARNINGS.getTFileUnsupportedKeys.code,
      Errors.GetTFileAttributes.InvalidDtoIn
    );

    //HDS 2
    const sysIdentitySession = await getSystemIdentity();
    let systemIdentityCode = sysIdentitySession.getClientIdentity().getAwid();

    // HDS 3
    // HDS 3.A
    // HDS 3.B
    if (!Config.get(Constants.TFileStoreUri)) {
      throw new Errors.GetTFileAttributes.TFileStoreUriConfigIsNotSet({ uuAppErrorMap });
    }

    // HDS 4
    let client = new AppClient({
      baseUri: Config.get(Constants.TFileStoreUri),
      session: sysIdentitySession,
    });

    const authorizedProfiles = await getAuthorizedProfiles(authzContext);

    let tFileObject;

    try {
      tFileObject = await client.get(Constants.Calls.getAttributes, { ...dtoIn, awid: systemIdentityCode });
    } catch (error) {
      // HDS 4.1
      throw new Errors.GetTFileAttributes.GetAttributesFailed(
        { uuAppErrorMap },
        {
          authorizedProfiles,
          uuConsoleBaseUri: error?.paramMap?.uuConsoleBaseUri,
          consoleCode: error?.paramMap?.consoleCode,
          progressBusCode: error?.paramMap?.progressBusCode,
        },
        error
      );
    }

    // HDS 5
    return { ...tFileObject, authorizedProfiles, uuAppErrorMap };
  }
}

module.exports = new GetTFileAttributesAbl();
