// uu_appstatisticsg01/src/components/constants/custom-stats-constants.js
"use strict";
//@@viewOn:revision
// coded: Eduard Batiushchenko (9108-9377-1)
// reviewed:
//@@viewOff:revision

//@@viewOn:constants
const Constants = {
  errors: {
    binaryGetDoesntExist: "uu-appbinarystore-main/binary/get/binaryDoesNotExist",
    consoleUriIsNotSet: "consoleUriIsNotSet",
  },
  defaultBinaryStore: "uu_app_binarystore_uri",
  binaryStoreBaseUri: "uu_app_statistics_binarystore_uri",
  scriptEngineBaseUri: "uu_app_statistics_script_engine_base_uri",
  scriptRepositoryUri: "uu_app_statistics_script_repository_base_uri",
  scriptCodesWhiteList: "uu_app_statistics_script_codes_whitelist",
  script: {
    get: "script/get",
  },
  collections: {
    management: "management",
    executives: "executives",
    uuAppStatistics: "uuAppStatistics",
  },
  useCases: {
    binary: {
      get: "binary/get",
      createAccess: "binary/createAccessKey",
      update: "binary/update",
      create: "binary/create",
    },
    scriptEngine: {
      runScript: "engine/runScript",
    },
  },
};
//@@viewOff:constants

//@@viewOn:exports
module.exports = Constants;
//@@viewOff:exports
